### PR TITLE
fix(tokenizer): treat special-token strings as text to prevent batch crashes (#3110)

### LINKED ISSUE #3110: ASR Inference Failure: Disallowed special<|no|>` causes single-sample transcription crash in batch processing
## 🐛 Bug

ASR transcription failed with special token validation error during batch dataset inference, while most samples run normally with stable RTF performance.

## To Reproduce

code: https://www.kaggle.com/code/liuweiq/funasr-recog-script?scriptVersionId=333049663

## Code sample

```python
    hotword = load_vocab(vocab_path)
    if hotword:
        print(f"已加载专业词汇：{hotword}")
    else:
        print("未找到词汇表文件，使用默认识别模式")
    
    print("\n正在加载模型...")
    model = AutoModel(
        model="FunAudioLLM/Fun-ASR-Nano-2512",
        hub="hf",
        trust_remote_code=True,
        vad_model="funasr/fsmn-vad",
        vad_kwargs={"max_single_segment_time": 30000},
        spk_model="funasr/campplus",
        device="cuda:0",
    )
    print("模型加载完成")
    
    audio_extensions = ('.mp3', '.wav', '.flac', '.ogg', '.m4a', '.MP3', '.WAV', '.FLAC')
    audio_files = [f for f in os.listdir(audio_dir) if f.endswith(audio_extensions)]
    
    if not audio_files:
        print("未找到音频文件")
        return
    
    print(f"\n找到 {len(audio_files)} 个音频文件")
    
    success_count = 0
    fail_count = 0
    
    for audio_file in audio_files:
        audio_path = os.path.join(audio_dir, audio_file)
        if transcribe_audio(audio_path, output_dir, hotword, model):
            success_count += 1
        else:
            fail_count += 1
```

## Expected behavior

no error
## Error logs

```text
<html>
<body>
<!--StartFragment-->
2441.9s | 18781 |  
-- | -- | --
2441.9s | 18782 |  
2441.9s | 18783 | {'load_data': '0.000', 'extract_feat': '0.007', 'forward': '0.652', 'batch_size': '1', 'rtf': '0.247'}, :  55%\|[34m█████▍    [0m\| 56/102 [00:33<00:28,  1.64it/s]
2441.9s | 18784 | {'load_data': '0.000', 'extract_feat': '0.007', 'forward': '0.652', 'batch_size': '1', 'rtf': '0.247'}, :  55%\|[34m█████▍    [0m\| 56/102 [00:33<00:28,  1.64it/s][A
2442.6s | 18785 | [A {'load_data': '0.000', 'extract_feat': '0.007', 'forward': '0.652', 'batch_size': '1', 'rtf': '0.247'}, :  55%\|[34m█████▍    [0m\| 56/102 [00:34<00:27,  1.65it/s]
2442.6s | 18786 | {'load_data': '0.000', 'extract_feat': '0.007', 'forward': '0.652', 'batch_size': '1', 'rtf': '0.247'}, :  55%\|[34m█████▍    [0m\| 56/102 [00:34<00:27,  1.65it/s]
2442.6s | 18787 | 0%\|[31m          [0m\| 0/1 [06:48<?, ?it/s]
2442.6s | 18788 |  
2442.7s | 18789 | 转录失败: Encountered text corresponding to disallowed special token '<\|no\|>'.
2442.7s | 18790 | If you want this text to be encoded as a special token, pass it to `allowed_special`, e.g. `allowed_special={'<\|no\|>', ...}`.
2442.7s | 18791 | If you want this text to be encoded as normal text, disable the check for this token by passing `disallowed_special=(enc.special_tokens_set - {'<\|no\|>'})`.
2442.7s | 18792 | To disable this check for all special tokens, pass `disallowed_special=()`.
2442.7s | 18793 |  

<br class="Apple-interchange-newline"><!--EndFragment-->
</body>
</html>
```

## Environment

- OS: Linux 6.12.90+
- Python version: 3.12.13
- FunASR version: 1.3.14
- ModelScope version: 1.38.0
- PyTorch / torchaudio version: 2.10.0+cu128 / 2.10.0+cu128
- Install method (pip, source, Docker): pip
- Device (cuda, cpu, mps): cuda
- GPU model: Tesla T4
- CUDA/cuDNN version: 12.8 / 91002
- Docker image tag, if used: gcr.io/kaggle-gpu-images/python@sha256:37c64f7dd9c54116ecd1bcc88817c5469b88387388fade02bfa8bf3fc647d461

## Audio details

If the audio cannot be shared, please describe:

Duration	3h 13m 38s（11,617s）
Sample rate	16,000 Hz
Format	MP3 (MPEG audio layer 3)
Bit rate	64 kbps
Channels	2 (stereo)
Sample format	32-bit float planar
File size	88.6 MB

### PR DESCRIPTION
**Pull request by 14790897**
## Summary

When transcribing long audio in a batch with `FunAudioLLM/Fun-ASR-Nano-2512` (+ `vad_model=fsmn-vad`, `spk_model=campplus`), a single sample can crash with `Encountered text corresponding to disallowed special token '<|no|>'.` and abort the entire batch, even though every other sample transcribes fine (stable RTF). Closes #3110.

**Root cause.** The model occasionally emits a special-token string (e.g. `<|no|>`, the Norwegian language tag) as part of its transcription text. That text is then re-encoded for forced alignment / loss computation:

```python
# funasr/models/fun_asr_nano/model.py:928
target_ids = torch.tensor(self.ctc_tokenizer.encode(result["text"]), dtype=torch.int64)
```

`self.ctc_tokenizer` resolves to `SenseVoiceTokenizer` (`funasr/tokenizer/whisper_tokenizer.py:30`) → `funasr/models/sense_voice/whisper_lib/tokenizer.py::Tokenizer`, whose `encode` forwards straight to `tiktoken.Encoding.encode`:

```python
def encode(self, text, **kwargs):
    return self.encoding.encode(text, **kwargs)
```

tiktoken defaults to `disallowed_special="all"`, so any text containing a special-token string raises `ValueError`, taking down the whole batch. `model.py:921` already strips `<|nospeech|>` from the CTC text — which confirms ASR output can leak special-token strings — but the allow-list isn't exhaustive (language tags like `<|no|>` slip through).

**Fix.** Default `disallowed_special=()` in `Tokenizer.encode` so special-token strings are treated as ordinary text unless a caller explicitly opts into stricter behaviour:

```python
def encode(self, text, **kwargs):
    kwargs.setdefault("disallowed_special", ())
    return self.encoding.encode(text, **kwargs)
```

## Type of change

- [x] Bug fix
- [ ] Documentation
- [ ] Example or demo
- [ ] Runtime or deployment
- [ ] Benchmark or evaluation
- [ ] Model/training change

## Validation

- [x] `python -m compileall funasr examples tests`
- [x] Docs or links checked
- [ ] Runtime/deployment command tested

`python -m compileall -q funasr examples tests` exits 0 (only two pre-existing `SyntaxWarning`s in `examples/industrial_data_pretraining/fun_asr_nano/tools/format5res.py`, untouched by this PR).

Adds `tests/test_sensevoice_tokenizer_special_tokens.py`. It builds a minimal in-memory `tiktoken.Encoding` (byte-level BPE + the specials `Tokenizer.__post_init__` needs), so it runs without the `multilingual.tiktoken` vocab (ships with the model, not this repo), without a GPU and without a model download:

- plain text encodes
- text containing `<|no|>` / `<|nospeech|>` no longer raises (regression for #3110)
- explicit `disallowed_special="all"` still raises (back-compat)
- `allowed_special="all"` still returns the special-token id (back-compat for `decoding.py`)

Local exercise of the fix against the minimal encoding (`importlib`-loaded `tokenizer.py` to bypass the heavy `funasr` package import chain on a machine without `torch`/`kaldiio`):

```
1. raw tiktoken encode("<|no|>")                        -> raises (reproduces #3110)
2. Tokenizer.encode("hello <|no|> world")               -> OK, 18 ids (fixed)
3. Tokenizer.encode("<|no|>", disallowed_special="all") -> still raises (back-compat)
4. Tokenizer.encode("<|no|>", allowed_special="all")   -> [257] == <|no|> id (back-compat)
```

## User impact

Production deployers running Fun-ASR-Nano (and any model reusing the SenseVoice `Tokenizer`) on batch / streaming ASR pipelines. Today a single bad sample aborts the whole batch; after this PR that sample is transcribed with the stray token treated as text and the batch completes. Researchers using `ctc_tokenizer.encode` for forced alignment / loss hit the same crash.

## Notes for reviewers

**Why the tokenizer, not the call sites.** There are multiple call sites that re-encode ASR output — `fun_asr_nano/model.py:443, 495, 923, 928`, `inference_vllm.py:668`, `inference_vllm_pipeline.py:348`. Patching the tokenizer covers all of them in one place, and also protects SenseVoice and any future model that reuses this `Tokenizer`.

**Backward compatibility.** `setdefault` keeps the change fully backward compatible:

- Callers that pass `disallowed_special=...` are untouched.
- `decoding.py`'s `tokenizer.encode(prompt, allowed_special="all")` still encodes special tokens as their token ids (verified in case 4 above).
- The internal `non_speech_tokens` cached_property calls `self.encoding.encode(...)` directly (not `self.encode`), so it's unaffected.

**Scope.** One-line behavioural default in `Tokenizer.encode` + regression tests. No model weights, no config, no public API change.

### CODE REVIEW

**Review by LauraGPT — APPROVED** [review]
Validated current head `1e42316c` on ind-gpu8:

- `python -m pytest tests/test_sensevoice_tokenizer_special_tokens.py -q` -> 6 passed
- `python -m py_compile funasr/models/sense_voice/whisper_lib/tokenizer.py tests/test_sensevoice_tokenizer_special_tokens.py` -> OK
- `git diff --check origin/main...HEAD` -> OK

No unresolved review threads are present, and the change directly fixes the batch crash from #3110 while preserving explicit `disallowed_special` / `allowed_special` caller behavior.
